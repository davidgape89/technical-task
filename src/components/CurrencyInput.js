import React from 'react';
import InputLabel from '@material-ui/core/InputLabel';
import FormControl from '@material-ui/core/FormControl';
import Input from '@material-ui/core/Input';
import InputAdornment from '@material-ui/core/InputAdornment';

const CURRENCY_REGEX = /^([0-9]{1,3},([0-9]{3},)*[0-9]{3}|[0-9]+).?([0-9]{0,2})$/;

export const CurrencyInput = ({
  label = 'Amount',
  value,
  placeholder,
  symbol,
  onChange,
}) => {
  const formatNumber = (number) => {
    number = number.toString();

    if (number && parseFloat(number) < Number.MAX_SAFE_INTEGER) {
      // Format integer part
      number = number.replace(/\d+/, integerPart => parseFloat(integerPart).toLocaleString('en-US'));
      // Take only two decimals
      return number.match(/[\d\,]*\.?\d{0,2}/)[0];
    }

    return '';
  };

  const onAmountChange = ({ target: { value } }) => {
    if (value.match(CURRENCY_REGEX)) {
      value = value.match(CURRENCY_REGEX)[0];
      value = value.replace(/\,/g, '');
      if (parseFloat(value) < Number.MAX_SAFE_INTEGER) {
        onChange(value);
      }
    } else if (value === '') {
      onChange('');
    }
  };

  return (
    <FormControl fullWidth>
      <InputLabel htmlFor="currency-input">{label}</InputLabel>
      <Input
        id="currency-input"
        variant="outlined"
        placeholder={placeholder}
        value={formatNumber(value)}
        onChange={onAmountChange}
        startAdornment={<InputAdornment position="start">{symbol}</InputAdornment>}
      />
    </FormControl>
  );
};
