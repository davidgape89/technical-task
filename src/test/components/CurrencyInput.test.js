import React from 'react';
import {shallow} from 'enzyme';

import Input from '@material-ui/core/Input';
import CurrencyInput from '../../components/CurrencyInput';

describe('CurrencyInput -', () => {
  let wrapper, onChange, defaultProps;

  beforeEach(() => {
    defaultProps = {
      label: 'Amount',
      value: '34',
      placeholder: '0',
      symbol: '$',
      onChange: onChange
    };
    onChange = jest.fn();
    wrapper = shallow(
      <CurrencyInput {...defaultProps} />
    );
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('number formatter -', () => {
    it('formats empty string correctly', () => {
      wrapper.setProps({ value: '' });

      expect(wrapper.find(Input).prop('value')).toBe('');
    });

    it('formats letters out', () => {
      wrapper.setProps({value: 'afds'});

      expect(wrapper.find(Input).prop('value')).toBe('');
    });

    it('formats short integers correctly', () => {
      wrapper.setProps({value: '22'});

      expect(wrapper.find(Input).prop('value')).toBe('22');
    });

    it('formats long integers correctly', () => {
      wrapper.setProps({value: '2222222222222'});

      expect(wrapper.find(Input).prop('value')).toBe('2,222,222,222,222');
    });

    it('does not accept numbers out of range', () => {
      wrapper.setProps({value: Number.MAX_SAFE_INTEGER.toString()});

      expect(wrapper.find(Input).prop('value')).toBe('');
    });

    it('formats numbers with decimals correctly', () => {
      wrapper.setProps({value: '22,222.22'});

      expect(wrapper.find(Input).prop('value')).toBe('22,222.22');
    });

    it('truncates down to two decimals', () => {
      wrapper.setProps({value: '22,222.2233432'});

      expect(wrapper.find(Input).prop('value')).toBe('22,222.22');
    });
  });

  describe('onAmountChange -', () => {
    let event = {
      target: {
        value: ''
      }
    };
    it('calls the function in the props correctly', () => {
      wrapper.setProps({onChange: onChange});
      wrapper.find(Input).prop('onChange')(event);

      expect(onChange).toHaveBeenCalled();
    });
  });
});