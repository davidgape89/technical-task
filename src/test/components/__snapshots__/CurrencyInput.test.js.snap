// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyInput - renders correctly 1`] = `
<WithStyles(FormControl)
  fullWidth={true}
>
  <WithStyles(WithFormControlContext(InputLabel))
    htmlFor="currency-input"
  >
    Amount
  </WithStyles(WithFormControlContext(InputLabel))>
  <WithStyles(Input)
    id="currency-input"
    onChange={[Function]}
    placeholder="0"
    startAdornment={
      <WithStyles(WithFormControlContext(InputAdornment))
        position="start"
      >
        $
      </WithStyles(WithFormControlContext(InputAdornment))>
    }
    value="34"
    variant="outlined"
  />
</WithStyles(FormControl)>
`;
